### PR TITLE
[SPARK-6638] [SQL] Improve performance of StringType in SQL

### PR DESCRIPTION
**Pull request by davies**
This PR change the internal representation for StringType from java.lang.String to UTF8String, which is implemented use ArrayByte.

This PR should not break any public API, Row.getString() will still return java.lang.String.

This is the first step of improve the performance of String in SQL.

cc @rxin

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SpecificMutableRow.scala [hunk 1]
```diff
@@ -230,13 +251,17 @@ final class SpecificMutableRow(val values: Array[MutableValue]) extends MutableR
...
   }
 
-  override def update(ordinal: Int, value: Any): Unit = {
-    if (value == null) setNullAt(ordinal) else values(ordinal).update(value)
+  override def update(ordinal: Int, value: Any): Unit = value match {
+    case null => setNullAt(ordinal)
+    case s: String =>
+      // for tests
+      throw new Exception("String should be converted into UTF8String")
+    case other => values(ordinal).update(value)
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SpecificMutableRow.scala]: I'm not actually sure, but I wonder if it would be faster to do a plain null check followed by an asInstanceOf, rather than pattern matching.  You will still get a class cast exception in tests if there is a mistake.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala [hunk 2]
```diff
@@ -258,11 +253,18 @@ case class Substring(str: Expression, pos: Expression, len: Expression) extends
...
       string match {
-        case ba: Array[Byte] => slice(ba, start, length)
-        case other => slice(other.toString, start, length)
+        case ba: Array[Byte] =>
+          val (st, end) = slicePos(start, length, () => ba.length)
+          ba.slice(st, end)
+        case s: UTF8String =>
+          val (st, end) = slicePos(start, length, () => s.length)
+          s.slice(st, end)
+        case other =>
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala]: Why do we need the `other` here still?  Could that indicate a bug elsewhere?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvaluationSuite.scala [hunk 3]
```diff
@@ -59,6 +60,10 @@ class ExpressionEvaluationBaseSuite extends FunSuite {
 
 class ExpressionEvaluationSuite extends ExpressionEvaluationBaseSuite {
 
+  def create_row(values: Array[Any]): Row = {
```
Comment (marmbrus) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvaluationSuite.scala]: Could we just use `Row.apply` here?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/GeneratedMutableEvaluationSuite.scala [hunk 4]
```diff
@@ -44,14 +43,6 @@ class GeneratedMutableEvaluationSuite extends ExpressionEvaluationSuite {
 
     val actual = plan(inputRow)
     val expectedRow = new GenericRow(Array[Any](expected))
-    if (actual.hashCode() != expectedRow.hashCode()) {
```
Comment (marmbrus) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/GeneratedMutableEvaluationSuite.scala]: Why removing this?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala (hunk 2 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala]: Yes, This case is added to match the old behavior, or some test will fail (other is Integer, maybe we does not check the dataType for substr?)

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/GeneratedMutableEvaluationSuite.scala (hunk 4 above)
Comment (davies) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/GeneratedMutableEvaluationSuite.scala]: The hash of UTF8String is different than String, this will fail, or will have to use UTF8String in expectedRow (updating all the tests). Which one do you like?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvaluationSuite.scala (hunk 3 above)
Comment (davies) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvaluationSuite.scala]: `Row.apply` will call ScalaRefaction.convertToCatalyst(), we still need a wrapper to do that.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/UTF8String.scala [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+ *
+ *  A String encoded in UTF-8 as an Array[Byte], which can be used for comparison,
+ *  search, see http://en.wikipedia.org/wiki/UTF-8 for details.
+ *
+ *  Note: This is not designed for general use cases, should not be used outside SQL.
+ */
+
+final class UTF8String extends Ordered[UTF8String] with Serializable {
+
+  private[this] var bytes: Array[Byte] = _
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/UTF8String.scala]: I had assumed that we would want to use `bytes: Array[Byte]` + `length: Int` so that the same byte array could be reused multiple times for different values.  It seems that allocating and zeroing out the byte arrays could actually be pretty expensive.

Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/UTF8String.scala]: Ah, okay talked to @rxin and we are going to try and do this later?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/GeneratedMutableEvaluationSuite.scala (hunk 4 above)
Comment (marmbrus) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/GeneratedMutableEvaluationSuite.scala]: I think its pretty important to make sure that all of our various internal row implementations have the same hashcode, as this has caused bugs in the past.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala (hunk 2 above)
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala]: Yeah, we probably need to add a rule for this in `FunctionArgumentConversion` to explicitly cast.  I assume this is a hive test?  I know we are pretty sloppy about this right now, but I think it would be best to start tightening up all of these cases as you find them.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/UTF8String.scala (hunk 5 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/UTF8String.scala]: Right now, `UTF8String` will take bytes from `Binary.getBytes` or `String.getBytes`, no copy, until we call copy() explicitly.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/GeneratedMutableEvaluationSuite.scala (hunk 4 above)
Comment (davies) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/GeneratedMutableEvaluationSuite.scala]: OK, I will add this back and fix the tests.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvaluationSuite.scala [hunk 6]
```diff
@@ -59,6 +60,10 @@ class ExpressionEvaluationBaseSuite extends FunSuite {
 
 class ExpressionEvaluationSuite extends ExpressionEvaluationBaseSuite {
 
+  def create_row(values: Any*): Row = {
+    new GenericRow(values.map(ScalaReflection.convertToCatalyst).toArray)
```
Comment (marmbrus) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvaluationSuite.scala]: I see.  Perhaps that is actually the way that we should handle inbound conversions for data coming from users, instead of doing it on inbound RDDs.  If we make `Row.apply` do the conversion then we always have converted data from users.  Low level APIs construct rows or mutable rows directly to avoid the conversion.

Comment (marmbrus) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvaluationSuite.scala]: Actually, maybe that is a bad idea... the problem is libraries like spark csv are using the internal rows so we would break compatibility...

Comment (davies) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvaluationSuite.scala]: Yes, that's also the thing I had considered. Once the Row is accessible from user, we can not use UTF8Sting inside it. For catalyst tests, we should do the conversion manually.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/Row.scala [hunk 7]
```diff
@@ -257,6 +257,7 @@ trait Row extends Serializable {
    *
    * @throws ClassCastException when data type does not match.
    */
+  // TODO(davies): This is not the right default implementation, we use Int as Date internally
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/Row.scala]: Are we excepting a `Date` object or its `Int` representation? For internal use, I guess we expect an int. But, since users can also call it, for them, a `Date` object is expected, right?
Right now, if I call `getDate`, I will get a `ClassCastException`?

Can you file a jira for it for 1.4 and mark it as a blocker?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala [hunk 8]
```diff
@@ -80,6 +82,24 @@ trait ScalaReflection {
     case (other, _) => other
   }
 
+  /**
+   *  Converts Scala objects to catalyst rows / types.
+   *  Note: This should be called before do evaluation on Row
+   *        (It does not support UDT)
+   */
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala]: It will be good to make it clear when we need this function.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 9]
```diff
@@ -323,7 +323,9 @@ case class Cast(child: Expression, dataType: DataType) extends UnaryExpression w
 
   private[this] def castToDecimal(from: DataType, target: DecimalType): Any => Any = from match {
     case StringType =>
-      buildCast[String](_, s => try changePrecision(Decimal(s.toDouble), target) catch {
+      buildCast[UTF8String](_, s => try {
+        changePrecision(Decimal(s.toString.toDouble), target)
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: Not quite related to your change. But, why we convert the string to double first?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SpecificMutableRow.scala [hunk 10]
```diff
@@ -202,6 +221,8 @@ final class SpecificMutableRow(val values: Array[MutableValue]) extends MutableR
         case DoubleType => new MutableDouble
         case BooleanType => new MutableBoolean
         case LongType => new MutableLong
+        // TODO(davies): enable this
+        // case StringType => new MutableString
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SpecificMutableRow.scala]: Why it is not enabled?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SpecificMutableRow.scala [hunk 11]
```diff
@@ -230,13 +251,19 @@ final class SpecificMutableRow(val values: Array[MutableValue]) extends MutableR
...
   }
 
-  override def update(ordinal: Int, value: Any): Unit = {
-    if (value == null) setNullAt(ordinal) else values(ordinal).update(value)
+  override def update(ordinal: Int, value: Any) {
+    if (value == null) {
+      setNullAt(ordinal)
+    } else {
+      // for tests
+      assert(!value.isInstanceOf[String], "String should be converted into UTF8String")
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SpecificMutableRow.scala]: Should we remove it?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SpecificMutableRow.scala [hunk 12]
```diff
@@ -230,13 +251,19 @@ final class SpecificMutableRow(val values: Array[MutableValue]) extends MutableR
...
+      values(ordinal).update(value)
+    }
   }
 
-  override def setString(ordinal: Int, value: String): Unit = update(ordinal, value)
+  override def setString(ordinal: Int, value: String): Unit = update(ordinal, UTF8String(value))
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SpecificMutableRow.scala]: Why we are still expecting a `String` as the input parameter?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 13]
```diff
@@ -216,10 +216,11 @@ abstract class CodeGenerator[InType <: AnyRef, OutType <: AnyRef] extends Loggin
           val $primitiveTerm: ${termForType(dataType)} = $value
          """.children
 
-      case expressions.Literal(value: String, dataType) =>
+      case expressions.Literal(value: UTF8String, dataType) =>
         q"""
           val $nullTerm = ${value == null}
-          val $primitiveTerm: ${termForType(dataType)} = $value
+          val $primitiveTerm: ${termForType(dataType)} =
+            org.apache.spark.sql.types.UTF8String(${value.toString})
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: I am not sure that I understand this line. We are getting the string representation of a `UTF8String` and then create a new `UTF8String`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 14]
```diff
@@ -595,6 +601,7 @@ abstract class CodeGenerator[InType <: AnyRef, OutType <: AnyRef] extends Loggin
       ordinal: Int,
       value: TermName) = {
     dataType match {
+      case StringType => q"$destinationRow.update($ordinal, $value)"
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: Seems this one is needed because `getString` returns a `String`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala [hunk 15]
```diff
@@ -85,8 +85,10 @@ case class UserDefinedGenerator(
   override protected def makeOutput(): Seq[Attribute] = schema
 
   override def eval(input: Row): TraversableOnce[Row] = {
+    // TODO(davies): improve this
+    val input_schema = StructType(children.map(e => StructField(e.simpleString, e.dataType, true)))
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala]: Can you explain what is the problem at here? Why we need to do the get the input schema in the `eval`? Also, `inputSchema` seems better than `input_schema`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala [hunk 16]
```diff
@@ -179,8 +179,7 @@ case class EqualTo(left: Expression, right: Expression) extends BinaryComparison
       val r = right.eval(input)
       if (r == null) null
       else if (left.dataType != BinaryType) l == r
-      else BinaryType.ordering.compare(
-        l.asInstanceOf[Array[Byte]], r.asInstanceOf[Array[Byte]]) == 0
+      else java.util.Arrays.equals(l.asInstanceOf[Array[Byte]], r.asInstanceOf[Array[Byte]])
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: Since you are touching the code of `EqualTo`, can you also fix the `EqualTo` in code gen (in `CodeGenerator.scala`)?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/DateUtils.scala [hunk 17]
```diff
@@ -39,6 +39,8 @@ object DateUtils {
     millisToDays(d.getTime)
   }
 
+  // TODO(davies): This is buggy, it will be wrong if the date is not aligned with day
+  // we should use the exact day as Int, for example, (year, month, day) -> day
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/DateUtils.scala]: Please open a jira for it. Also, add an example to explain the problem if you think the `TODO` at here is necessary.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/UTF8String.scala (hunk 5 above)
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/UTF8String.scala]: What is `Binary` that you are referring at here? Also, can you explain what do you mean by `no copy` at here?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/UTF8String.scala [hunk 18]
```diff
@@ -0,0 +1,209 @@
...
+  /**
+   * Update the UTF8String with Array[Byte], which should be encoded in UTF-8
+   */
+  def set(bytes: Array[Byte]): UTF8String = {
+    this.bytes = bytes
+    this
+  }
+
+  @inline
+  private[this] def numOfBytes(b: Byte): Int = {
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/UTF8String.scala]: Maybe add a comment at here to explain it?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/UTF8String.scala [hunk 19]
```diff
@@ -0,0 +1,209 @@
...
+  override def compareTo(other: UTF8String): Int = {
+    compare(other)
+  }
+
+  override def equals(other: Any): Boolean = other match {
+    case s: UTF8String =>
+      Arrays.equals(bytes, s.getBytes)
+    case s: String =>
+      // fail fast
+      bytes.length >= s.length && length() == s.length && toString() == s
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/UTF8String.scala]: When do we need this?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ExistingRDD.scala [hunk 20]
```diff
@@ -54,6 +52,27 @@ object RDDConversions {
       }
     }
   }
+  def rowToRowRdd(data: RDD[Row], schema: StructType): RDD[Row] = {
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ExistingRDD.scala]: Add a comment to explain the reason that we need it?

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JDBCRDD.scala [hunk 21]
```diff
@@ -349,12 +349,14 @@ private[sql] class JDBCRDD(
           val pos = i + 1
           conversions(i) match {
             case BooleanConversion    => mutableRow.setBoolean(i, rs.getBoolean(pos))
+            // TODO(davies): convert Date into Int
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/JDBCRDD.scala]: Since the `JvmType` of `DateType` is `Int`, I am wondering why we are still handle `java.sql.Date` internally. Is there a jira for it?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/UTF8String.scala (hunk 19 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/UTF8String.scala]: For tests.

File: sql/core/src/test/java/test/org/apache/spark/sql/JavaRowSuite.java [hunk 22]
```diff
@@ -17,6 +17,12 @@
 
 package test.org.apache.spark.sql;
 
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.RowFactory;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
```
Comment (yhuai) [inline sql/core/src/test/java/test/org/apache/spark/sql/JavaRowSuite.java]: Unnecessary change?

File: sql/hive/v0.13.1/src/main/scala/org/apache/spark/sql/hive/Shim13.scala [hunk 23]
```diff
@@ -224,7 +218,7 @@ private[hive] object HiveShim {
       TypeInfoFactory.voidTypeInfo, null)
 
   def getStringWritable(value: Any): hadoopIo.Text =
-    if (value == null) null else new hadoopIo.Text(value.asInstanceOf[String])
+    if (value == null) null else new hadoopIo.Text(value.asInstanceOf[UTF8String].toString)
```
Comment (yhuai) [inline sql/hive/v0.13.1/src/main/scala/org/apache/spark/sql/hive/Shim13.scala]: Since you are changing `Shim13`, do we need to update `Shim12` as well?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/UTF8String.scala (hunk 19 above)
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/UTF8String.scala]: So, we do not expect that `other` is a `String` in a real use case, right? If so, why not change the test and get rid of it?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/UTF8String.scala (hunk 5 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/UTF8String.scala]: The `Binary` is `parquet.io.api.Binary`, When we create a `UTFString` from `Binary.getBytes`, we does not need to do another copy for bytes.

Before this patch, we will create a copy as `String`.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JDBCRDD.scala (hunk 21 above)
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/JDBCRDD.scala]: Created: https://issues.apache.org/jira/browse/SPARK-6784

File: sql/catalyst/src/main/scala/org/apache/spark/sql/Row.scala (hunk 7 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/Row.scala]: https://issues.apache.org/jira/browse/SPARK-6784

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala (hunk 9 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: I guess Double is the wide format and range for numbers, or we need to have a special parser for it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SpecificMutableRow.scala (hunk 10 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SpecificMutableRow.scala]: We need more thinking about reuse the UTF8String object, it's not a trivial decision, so I'd like to leave this out of this PR.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SpecificMutableRow.scala (hunk 12 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SpecificMutableRow.scala]: This is public API, so we can not change it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/DateUtils.scala (hunk 17 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/DateUtils.scala]: https://issues.apache.org/jira/browse/SPARK-6785

File: sql/hive/v0.13.1/src/main/scala/org/apache/spark/sql/hive/Shim13.scala (hunk 23 above)
Comment (davies) [inline sql/hive/v0.13.1/src/main/scala/org/apache/spark/sql/hive/Shim13.scala]: Good catch, done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala (hunk 16 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala (hunk 14 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: yes.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/UTF8String.scala (hunk 19 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/UTF8String.scala]: I tried to remove this, then we need to convert String into UTF8String manually in many cases, especially, a String inside a Map/Struct/Array.

This could simply the tests a lot, I'd like to keep this.

**Comment by davies** [comment]
@yhuai I should had addressed all you comments, or leave a comment there, could you take another look?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateProjection.scala [hunk 24]
```diff
@@ -111,36 +111,54 @@ object GenerateProjection extends CodeGenerator[Seq[Expression], Projection] {
 
     val specificAccessorFunctions = NativeType.all.map { dataType =>
       val ifStatements = expressions.zipWithIndex.flatMap {
-        case (e, i) if e.dataType == dataType =>
+        // getString() is not used by expressions
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateProjection.scala]: Explain it a little bit more.

**Comment by davies** [comment]
@marmbrus Could you take a final look on this?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateProjection.scala (hunk 24 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateProjection.scala]: done, see comments below.

File: sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala [hunk 25]
```diff
@@ -126,6 +126,13 @@ abstract class BaseRelation {
    * could lead to execution plans that are suboptimal (i.e. broadcasting a very large table).
    */
   def sizeInBytes: Long = sqlContext.conf.defaultSizeInBytes
+
+  /**
+   * Whether does it need to convert the objects in Row to internal representation, for example:
+   *  java.lang.String -> UTF8String
+   *  java.lang.Decimal -> Decimal
+   */
+  def needConversion: Boolean = true
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: We should comment that the internal representation is not stable across releases and thus data sources outside of Spark SQL should leave this as true.

File: sql/core/src/main/scala/org/apache/spark/sql/sources/DataSourceStrategy.scala [hunk 26]
```diff
@@ -102,20 +102,29 @@ private[sql] object DataSourceStrategy extends Strategy {
...
-        execution.PhysicalRDD(requestedColumns, scanBuilder(requestedColumns, pushedFilters))
+      val scan = createPhysicalRDD(relation.relation, requestedColumns,
+        scanBuilder(requestedColumns, pushedFilters))
       execution.Project(projectList, filterCondition.map(execution.Filter(_, scan)).getOrElse(scan))
     }
   }
 
+  private[this] def createPhysicalRDD(relation: BaseRelation,
+                                      output: Seq[Attribute],
+                                      rdd: RDD[Row]) = {
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/DataSourceStrategy.scala]: indent 4 spaces on next line.

**Comment by davies** [comment]
@marmbrus done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystTypeConverters.scala [hunk 27]
```diff
@@ -284,9 +321,9 @@ object CatalystTypeConverters {
       row: Row,
       schema: StructType,
       converters: Array[Any => Any]): Row = {
-    val ar = new Array[Any](row.size)
+    val ar = new Array[Any](converters.size)
     var idx = 0
-    while (idx < row.size) {
+    while (idx < converters.size && idx < row.size) {
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystTypeConverters.scala]: Why the number of converters is different with the number of fields in a row?

Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystTypeConverters.scala]: For HiveQuerySuite "Add JAR command 2", the size of row is 1, but the number of fields is 0, maybe Hive 1.3 has changed the result?

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystTypeConverters.scala]: can we change size to length? otherwise it allocates a new object for arrays

Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystTypeConverters.scala]: btw, where is "ADD JAR command 2"? I could not find it...

**Comment by rxin** [comment]
BTW - since this changes so many files, it'd be great to merge this as soon as possible. We can fix minor problems later in follow up PRs.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystTypeConverters.scala (hunk 27 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystTypeConverters.scala]: It's a new test case in master, I had to merged with master to debug it.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/commands.scala [hunk 28]
```diff
@@ -76,6 +76,12 @@ case class DropTable(
 private[hive]
 case class AddJar(path: String) extends RunnableCommand {
 
+  override val output: Seq[Attribute] = {
+    val schema = StructType(
+      StructField("result", IntegerType, false) :: Nil)
+    schema.toAttributes
+  }
```
Comment (yhuai) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/commands.scala]: I do not really know the reason that the result of AddJar is a `Row(0)` (see a few lines below.). But, we can figure it out after we merge it.

Comment (yhuai) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/commands.scala]: OK, the reason is to match the behavior of Hive... This change looks good.

**Comment by yhuai** [comment]
LGTM

**Comment by marmbrus** [comment]
Thanks!  Merged to master!
